haskellPackages.mkDerivation: Replace testTarget with testTargets

TeofilC opened the pull request:
## Description of changes

testTarget takes a space separated list of test suites to run. We introduce a new testTargets argument that instead takes a list, and add a backwards compatibility shim.

cc @sternenseemann  you mentioned you were interested in a change like this.

I've tested building a couple of packages and it seems to evaluate to the same thing. 

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

maralorn commented:
I like this change.

Is there a way to actually throw a warning if someone explicitly sets the old value?

TeofilC commented:
> Is there a way to actually throw a warning if someone explicitly sets the old value?

This sounds good. I'll make it emit a warning (using `lib.warn`) if a user tries to use the old setting.

TeofilC committed:
haskellPackages.mkDerivation: Replace testTarget with testTargets

testTarget takes a space separated list of test suites to run. We itnroduce a new testTargets argument that instead takes a list, and add a backwards compatibility shim

TeofilC commented:
Sorry for this taking so long. I've now added a warning @maralorn. 
It looks like `cabal2nix` has some hard-coded overrides that always generate `testTarget` attrs for some packages, so we'd have to change that to avoid the deprecation message with something like https://github.com/NixOS/cabal2nix/pull/646

sternenseemann commented on pkgs/development/haskell-modules/configuration-nix.nix:
```diff
@@ -701,7 +701,7 @@ self: super: builtins.intersectAttrs super {
   # Not running the "example" test because it requires a binary from lsps test
   # suite which is not part of the output of lsp.
-  lsp-test = overrideCabal (old: { testTarget = "tests func-test"; }) super.lsp-test;
+  lsp-test = overrideCabal (old: { testTargets = [ "tests func-test" ]; }) super.lsp-test;
```
```suggestion
  lsp-test = overrideCabal (old: { testTargets = [ "tests" "func-test" ]; }) super.lsp-test;
```

sternenseemann commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -408,6 +410,10 @@ let
     '';
 
+  testTargetsString =
+    lib.warnIf (testTarget != "") "haskell-modules/generic-builder.testTarget is deprecated. Use testTargets instead"
```
```suggestion
    lib.warnIf (testTarget != "") "haskellPackages.mkDerivation: testTarget is deprecated. Use testTargets instead"
```

sternenseemann commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -408,6 +410,10 @@ let
 
+  testTargetsString =
+    lib.warnIf (testTarget != "") "haskell-modules/generic-builder.testTarget is deprecated. Use testTargets instead"
+    (lib.concatStringsSep " " testTargets);
```
In a follow up change, we could change this to `lib.escapeShellArgs` (which of course causes a rebuild).

sternenseemann reviewed (commented):
This looks great, thanks!

- Needs a changelog entry for NixOS 25.05
- Please add a `haskellPackages.mkDerivation: ` prefix to the commit message.

sternenseemann commented:
https://github.com/NixOS/nixpkgs/pull/381187 implements a compatible interface for 24.11 so expressions can work with both unstable/25.05 and 24.11.
